Update subdomain configs

brgeen opened the pull request:
Different DNS endpoints for internal switchboard communication [Task](https://app.asana.com/0/1206395725449160/1207327192512421/f)

### Changes

This moves all host, domain, subdomain, and HTTP configs to the Arux App Gem. 

1. Provides the top level configuration and
logic to take a domain, subdomain and create an
HTTPS URI with a Public or API hostname.
2. Each of these are environment and application
specific and set in the Arux App Gem.

This allows us to direct which subdomain we would like 
the traffic routed through.

### Test
1. Dependent on: 
https://github.com/Arux-Software/dev_setup/pull/47
https://github.com/Arux-Software/TCE/pull/11890
2. Login through the TCE community portal
- Add a course to your cart
- Checkout with a test card
3. Login through the TCE admin portal
- Add a course to the cart
- Checkout via admin cart
4. Call any of the `Switchboard::SwitchboardConnection` modules in TCE
```ruby
Switchboard::AccountsConnection.get_user_data(User.last.uuid)
```
5. Call any of the library methods directly
```ruby
connection = AruxApp::API::Account.new(
  auth: Switchboard::SwitchboardConnection.auth_config,
  access_token: Switchboard::SwitchboardConnection.client_credentials_token
)
connection.get(User.last.uuid)
```

```ruby
AruxApp::API::Account.api_uri
```

brgeen committed:
move dev HOSTNAME initialization to top

This is now being used in a configuration hash and
needs to be initialized before the modules are
required.

brgeen committed:
add and require checkout module

This adds a new module for the Checkout app and
includes the HTTPS configs necessary to connect to
the application.

This also includes URL's for a necessary tokenx
iframe

brgeen committed:
remove nav and student modules

These do not appear to be used by any of our
applications.

brgeen committed:
build out top-level domains hash and methods

This provides the top level configuration and
logic to take an environment, and combine an
environment specific domain name with a module
specific subdomain to create an HTTPS URI with a
public or API hostname.

brgeen committed:
replace .server_uri with .public_uri and .api_uri

This allows us to set a specific subdomain for
each module

brgeen committed:
replace .server_uri usages

This replaces all usages of the old method with
either .public_uri or .api_uri depending on the
subdomain we want to route the traffic through

brgeen committed:
remove user lock methods

User locks have been removed as part of Arux
Access, this logic is no longer used.

brgeen committed:
update semantic version to 3.0.0

The reason for the major change; this will cause
breaking changes to any application that updates
versions but does not update usages in
application logic.

chriswarren commented on lib/arux_app/api.rb:
```diff
@@ -1,29 +1,27 @@
-        end
+    class << self
+      def mode
+        raise "Environment is not set, i.e. ARUX_APP_GEM_MODE = :development" unless const_defined?(:ARUX_APP_GEM_MODE)
```
Any reason not to do the same check as in `domain`? And if we do it here, I think we could remove line 23 too

```suggestion
        raise "Environment is not set to a valid environment, i.e. ARUX_APP_GEM_MODE = :development" if !DOMAINS.has_key?(ARUX_APP_GEM_MODE)
```

brgeen replied:
I played around with that, and it appears we need the const_defined method while passing in a symbol. Otherwise we get something like this
```
        5: from /Users/brooksgeenen/arux/arux_app_gem/lib/arux_app/api.rb:12:in `mode'
        4: from /Users/brooksgeenen/.rvm/gems/ruby-2.5.0-nrbgem18@tce-v2/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:120:in `const_missing'
        3: from /Users/brooksgeenen/.rvm/gems/ruby-2.5.0-nrbgem18@tce-v2/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:130:in `rescue in const_missing'
        2: from /Users/brooksgeenen/.rvm/gems/ruby-2.5.0-nrbgem18@tce-v2/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:417:in `qualified_name_for'
        1: from /Users/brooksgeenen/.rvm/gems/ruby-2.5.0-nrbgem18@tce-v2/gems/activesupport-2.3.18/lib/active_support/dependencies.rb:611:in `to_constant_name'
ArgumentError (Anonymous modules have no name to be referenced by)
```

chriswarren replied:
`const_get(:ARUX_APP_GEM)` maybe?

Or check for it to be defined, but still check if it's a valid key in `DOMAINS`?

Or leave it as-is; what's here is fine too.

brgeen replied:
> `const_get(:ARUX_APP_GEM)` maybe?

Hmm, still throws the same error, I think I might leave as is for now, since they are both working in tandem and checking each other.

chriswarren commented on lib/arux_app/api.rb:
```diff
@@ -1,29 +1,27 @@
-          "https://account.#{HOSTNAME}"
-        end
+      def domain
+        raise "#{mode} is not a supported environment" unless DOMAINS.has_key?(mode)
```
Don't need this if we put it in `mode`
```suggestion
```

brgeen replied:
Could be wrong, but I think we need both of them, based on the comment above. Happy to explore more and pair through it if you'd like.

chriswarren commented on lib/arux_app/api/account.rb:
```diff
@@ -207,7 +155,7 @@ def delete_relationship
 
       def api_route
-        "#{self.class.server_uri}/api/v#{api_version}"
+        "#{self.class.api_uri}/api/v#{api_version}"
```
I wonder if we should do instance methods that handle `self.class` since we're stepping through it in this way quite frequently

brgeen replied:
Thats a good call. 

Do you think every module should have its uri methods setup like this (for consistency), even if the module is not currently using an instance method?

```
      def self.public_uri
        AruxApp::API.uri(subdomain: "cart")
      end

      def public_uri
        self.class.public_uri
      end
      
      def self.api_uri
        AruxApp::API.uri(subdomain: "cart.api")
      end

      def api_uri
        self.class.api_uri
      end
```

chriswarren replied:
I'm good either way, I think

Typically would avoid putting unused methods in just in case, but in a gem it probably makes a little more sense - whatever you pick would work for me

chriswarren reviewed (approved):
looks good - a couple ideas, if you like

brgeen committed:
define URI instance methods for public and api

This allows us to keep things DRY instead of using
 self.class.uri whenever self is an instance.

nhaglind reviewed (commented):
Should we update the `arux.app.gemspec` to bump the version?

brgeen committed:
bump gemspec version
